File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        PriorityQueue<ListNode> pq = new PriorityQueue<>(lists.length, (a,b) -> a.val - b.val);
        ListNode dummy = new ListNode(), p = dummy;
        for (ListNode node : lists) {
            if (node != null) {
                pq.add(node);
            }
        }
        while (!pq.isEmpty()) {
            ListNode min = pq.poll();
            p.next = min;
            min = min.next;
            if (min != null) pq.add(min);
            p = p.next;
        }
        return dummy.next;
    }
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[Optional[ListNode]]
        :rtype: Optional[ListNode]
        """
        pq = []
        for lst in lists:
            if lst:
                heapq.heappush(pq, (lst.val, lst))
        dummy = ListNode()
        p = dummy
        while pq:
            val, lst = heapq.heappop(pq)
            p.next = lst
            p = p.next
            lst = lst.next
            if lst:
                heapq.heappush(pq, (lst.val, lst))
        return dummy.next


        